File: Algorithms/Data Structure dependent/Arrays and String/Sorted Arrays/Binary Search/threeSum.java
class Solution {
    public List<List<Integer>> threeSum(int[] nums){
        List<List<Integer>> res = new LinkedList<>();
        Arrays.sort(nums);
        for(int i=0; i<nums.length-2; i++){
            if(i>0 && nums[i] == nums[i-1]) continue;
            twoSums(nums, i, res);
        }
        return res;
    }

    public void twoSums(int[] nums, int index, List<List<Integer>> res){
        int l=index+1, r=nums.length-1;
        int target = -nums[index];
        while(l<r){
            int curSum = nums[l] + nums[r];
            if(curSum == target){
                res.add(Arrays.asList(-target, nums[l],nums[r]));
                /* Skip duplicates */
                while(l<r && nums[l] == nums[l+1]) l++;
                while(l<r && nums[r] == nums[r-1]) r--;
            }
            if(curSum <= target){
                l++;
            }
            if(curSum>=target){
                r--;
            } 
        }
    }
}

File: Algorithms/Data Structure dependent/Arrays and String/Two Pointers/threeSum.go
import
(
    "sort"
)
func threeSum(nums []int) [][]int {
    sort.Ints(nums)
    const target = 0
    out := make([][]int, 0)
    for i, x := range nums {
        if i > 0 && nums[i-1] == x{
            continue
        }
        curSums := twoSums(nums, i+1, target-x)
        for _, pair := range curSums {
            curPair := []int{x}
            out = append(out, append(curPair, pair...))
        }
    }

    return out
}

func twoSums(nums []int, start int, target int) [][]int{
    end := len(nums)-1
    l, r := start, end
    out := make([][]int, 0)
    for l<r {
        curSum := nums[l] + nums[r]
        if curSum < target || (l > start && nums[l] == nums[l-1]){
            l++
        } else if curSum > target || (r+1 < end && nums[r] == nums[r+1]) {
            r--
        } else {
            curPair := []int{nums[l], nums[r]}
            out = append(out, curPair)
            l++
            r--
        }
    }
    return out
}